File: SlidingWindowandTwoPointers/Longest Festival Streak/manasvi9745_soln.py
def longest_streak_optimal(a, k):
    n = len(a)
    left = 0
    current_sum = 0
    max_length = 0

    for right in range(n):
        current_sum += a[right]

        while current_sum > k:
            current_sum -= a[left]
            left += 1

        max_length = max(max_length, right - left + 1)

    return max_length


# Example
a = [2, 1, 3, 2, 1]
k = 7
print(longest_streak_optimal(a, k))  # 3

File: SlidingWindowandTwoPointers/Window of Distinct Relics/manasvi9745_soln.java
import java.util.*;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int n = sc.nextInt();
        int k = sc.nextInt();
        int[] arr = new int[n];

        for (int i = 0; i < n; i++)
            arr[i] = sc.nextInt();

        HashMap<Integer, Integer> freq = new HashMap<>();

        int left = 0;
        int maxLength = 0;

        for (int right = 0; right < n; right++) {
            freq.put(arr[right], freq.getOrDefault(arr[right], 0) + 1);

            while (freq.size() > k) {
                freq.put(arr[left], freq.get(arr[left]) - 1);

                if (freq.get(arr[left]) == 0)
                    freq.remove(arr[left]);

                left++;
            }

            maxLength = Math.max(maxLength, right - left + 1);
        }

        System.out.println(maxLength);
    }
}